core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -111,31 +172,39 @@ private[fs2] object Algebra {
       g: (B, O) => B,
       v: FreeC[Algebra[F, O, ?], Unit]
   )(implicit F: Sync[F]): F[B] = {
-    type UnconsF[X] =
-      Function2[
-        CompileScope[F, O],
-        Either[Throwable, Option[(Segment[X, Unit], FreeC[Algebra[F, X, ?], Unit])]],
-        F[(CompileScope[F, O], FreeC[Algebra[F, O, ?], Unit])]
-      ]
+
+    // Uncons is interrupted, fallback on `compileFoldLoop` has to be invoked
+    // The token is a scope from which we should recover.
+    case class Interrupted[X](token: Token) extends UR[X]
```

Comment:
Can we move these classes to directly under `Algebra` to ensure they don't accidentally close over any state from this method?

Reply:
Oh I guess the close over `F`

Reply:
Well, there is unfortunatelly disrepance (see Continue) between ones used in compile and translate. In fact closing them inside F/G allowed me to discover another typecast bug. 

Having said that, once we will battle-test this implementation, we could mover them outside compile//translate